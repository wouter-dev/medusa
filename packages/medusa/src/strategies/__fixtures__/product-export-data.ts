import { IdMap } from "medusa-test-utils"

export const productsToExport = [
  {
    collection: {
      created_at: "randomString",
      deleted_at: null,
      handle: "test-collection1",
      id: IdMap.getId("product-export-collection_1"),
      metadata: null,
      title: "Test collection 1",
      updated_at: "randomString",
    },
    collection_id: IdMap.getId("product-export-collection_1"),
    created_at: "randomString",
    deleted_at: null,
    description: "test-product-description-1",
    discountable: true,
    external_id: null,
    handle: "test-product-product-1",
    height: null,
    hs_code: null,
    id: IdMap.getId("product-export-product_1"),
    images: [
      {
        created_at: "randomString",
        deleted_at: null,
        id: IdMap.getId("product-export-image_1"),
        metadata: null,
        updated_at: "randomString",
        url: "test-image.png",
      },
    ],
    is_giftcard: false,
    length: null,
    material: null,
    metadata: null,
    mid_code: null,
    options: [
      {
        created_at: "randomString",
        deleted_at: null,
        id: IdMap.getId("product-export-option_1"),
        metadata: null,
        product_id: IdMap.getId("product-export-product_1"),
        title: "test-option-1",
        updated_at: "randomString",
      },
      {
        created_at: "randomString2",
        deleted_at: null,
        id: IdMap.getId("product-export-option_2"),
        metadata: null,
        product_id: IdMap.getId("product-export-product_1"),
        title: "test-option-2",
        updated_at: "randomString2",
      },
    ],
    origin_country: null,
    profile_id: IdMap.getId("product-export-profile_1"),
    profile: {
      id: IdMap.getId("product-export-profile_1"),
      name: "profile_1",
      type: "profile_type_1",
    },
    status: "draft",
    subtitle: null,
    tags: [
      {
        created_at: "randomString",
        deleted_at: null,
        id: IdMap.getId("product-export-tag_1"),
        metadata: null,
        updated_at: "randomString",
        value: "123_1",
      },
    ],
    thumbnail: null,
    title: "Test product",
    type: {
      created_at: "randomString",
      deleted_at: null,
      id: IdMap.getId("product-export-type_1"),
      metadata: null,
      updated_at: "randomString",
      value: "test-type-1",
    },
    type_id: IdMap.getId("product-export-type_1"),
    updated_at: "randomString",
    variants: [
      {
        allow_backorder: false,
        barcode: "test-barcode",
        calculated_price: null,
        created_at: "randomString",
        deleted_at: null,
        ean: "test-ean",
        height: null,
        hs_code: null,
        id: IdMap.getId("product-export-variant_1"),
        inventory_quantity: 10,
        length: null,
        manage_inventory: true,
        material: null,
        metadata: null,
        mid_code: null,
        options: [
          {
            created_at: "randomString",
            deleted_at: null,
            id: IdMap.getId("product-export-variant_option_1"),
            metadata: null,
            option_id: IdMap.getId("product-export-option_1"),
            updated_at: "randomString",
            value: "option 1 value 1",
            variant_id: IdMap.getId("product-export-variant_1"),
          },
          {
            created_at: "randomString",
            deleted_at: null,
            id: IdMap.getId("product-export-variant_option_2"),
            metadata: null,
            option_id: IdMap.getId("product-export-option_2"),
            updated_at: "randomString",
            value: "option 2 value 1",
            variant_id: IdMap.getId("product-export-variant_1"),
          },
        ],
        origin_country: null,
        original_price: null,
        prices: [
          {
            amount: 100,
            created_at: "randomString",
            currency_code: "usd",
            deleted_at: null,
            id: IdMap.getId("product-export-price_1"),
            max_quantity: null,
            min_quantity: null,
            price_list: null,
            price_list_id: null,
            region_id: null,
            updated_at: "randomString",
            variant_id: IdMap.getId("product-export-variant_1"),
          },
        ],
        product_id: IdMap.getId("product-export-product_1"),
        sku: "test-sku",
        title: "Test variant",
        upc: "test-upc",
        updated_at: "randomString",
        weight: null,
        width: null,
      },
    ],
    weight: null,
    width: null,
  },
  {
    collection: {
      created_at: "randomString",
      deleted_at: null,
      handle: "test-collection2",
      id: IdMap.getId("product-export-collection_2"),
      metadata: null,
      title: "Test collection",
      updated_at: "randomString",
    },
    collection_id: "test-collection",
    created_at: "randomString",
    deleted_at: null,
    description: "test-product-description",
    discountable: true,
    external_id: null,
    handle: "test-product-product-2",
    height: null,
    hs_code: null,
    id: IdMap.getId("product-export-product_2"),
    images: [
      {
        created_at: "randomString",
        deleted_at: null,
        id: IdMap.getId("product-export-image_2"),
        metadata: null,
        updated_at: "randomString",
        url: "test-image.png",
      },
    ],
    is_giftcard: false,
    length: null,
    material: null,
    metadata: null,
    mid_code: null,
    options: [
      {
        created_at: "randomString",
        deleted_at: null,
        id: IdMap.getId("product-export-option_2"),
        metadata: null,
        product_id: IdMap.getId("product-export-product_2"),
        title: "test-option",
        updated_at: "randomString",
      },
    ],
    origin_country: null,
    profile_id: IdMap.getId("product-export-profile_2"),
    profile: {
      id: IdMap.getId("product-export-profile_2"),
      name: "profile_2",
      type: "profile_type_2",
    },
    status: "draft",
    subtitle: null,
    tags: [
      {
        created_at: "randomString",
        deleted_at: null,
        id: IdMap.getId("product-export-tag_2"),
        metadata: null,
        updated_at: "randomString",
        value: "123",
      },
    ],
    thumbnail: null,
    title: "Test product",
    type: {
      created_at: "randomString",
      deleted_at: null,
      id: IdMap.getId("product-export-type_2"),
      metadata: null,
      updated_at: "randomString",
      value: "test-type",
    },
    type_id: "test-type",
    updated_at: "randomString",
    variants: [
      {
        allow_backorder: false,
        barcode: "test-barcode",
        calculated_price: null,
        created_at: "randomString",
        deleted_at: null,
        ean: "test-ean",
        height: null,
        hs_code: null,
        id: IdMap.getId("product-export-variant_2"),
        inventory_quantity: 10,
        length: null,
        manage_inventory: true,
        material: null,
        metadata: null,
        mid_code: null,
        options: [
          {
            created_at: "randomString",
            deleted_at: null,
            id: IdMap.getId("product-export-variant_option_2"),
            metadata: null,
            option_id: IdMap.getId("product-export-option_2"),
            updated_at: "randomString",
            value: "Option 1 value 1",
            variant_id: IdMap.getId("product-export-variant_2"),
          },
        ],
        origin_country: null,
        original_price: null,
        prices: [
          {
            amount: 100,
            created_at: "randomString",
            currency_code: "usd",
            deleted_at: null,
            id: IdMap.getId("product-export-price_2"),
            max_quantity: null,
            min_quantity: null,
            price_list: null,
            price_list_id: null,
            region_id: null,
            updated_at: "randomString",
            variant_id: IdMap.getId("product-export-variant_2"),
          },
        ],
        product_id: IdMap.getId("product-export-product_2"),
        sku: "test-sku",
        title: "Test variant",
        upc: "test-upc",
        updated_at: "randomString",
        weight: null,
        width: null,
      },
      {
        allow_backorder: false,
        barcode: "test-barcode",
        calculated_price: null,
        created_at: "randomString",
        deleted_at: null,
        ean: "test-ean",
        height: null,
        hs_code: null,
        id: IdMap.getId("product-export-variant_3"),
        inventory_quantity: 10,
        length: null,
        manage_inventory: true,
        material: null,
        metadata: null,
        mid_code: null,
        options: [
          {
            created_at: "randomString",
            deleted_at: null,
            id: IdMap.getId("product-export-variant_option_2"),
            metadata: null,
            option_id: IdMap.getId("product-export-option_2"),
            updated_at: "randomString",
            value: "Option 1 Value 1",
            variant_id: IdMap.getId("product-export-variant_3"),
          },
        ],
        origin_country: null,
        original_price: null,
        prices: [
          {
            amount: 100,
            created_at: "randomString",
            currency_code: "usd",
            deleted_at: null,
            id: IdMap.getId("product-export-price_2"),
            max_quantity: null,
            min_quantity: null,
            price_list: null,
            price_list_id: null,
            region_id: null,
            updated_at: "randomString",
            variant_id: IdMap.getId("product-export-variant_3"),
          },
        ],
        product_id: IdMap.getId("product-export-product_2"),
        sku: "test-sku",
        title: "Test variant",
        upc: "test-upc",
        updated_at: "randomString",
        weight: null,
        width: null,
      },
    ],
    weight: null,
    width: null,
  },
]
