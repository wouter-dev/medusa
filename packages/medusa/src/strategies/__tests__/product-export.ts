import ProductExportStrategy from "../product-export"
import { IdMap, MockManager, MockRepository } from "medusa-test-utils"
import { User } from "../../models"
import { BatchJobStatus } from "../../types/batch-job"
import { productsToExport } from "../__fixtures__/product-export-data"

const outputDataStorage: string[] = []

let fakeJob = {
  id: IdMap.getId("product-export-job"),
  type: 'product-export',
  context: {
    listConfig: {},
    filterableFields: {},
  },
  created_by: IdMap.getId("product-export-job-creator"),
  created_by_user: {} as User,
  result: {},
  dry_run: false,
  status: BatchJobStatus.PROCESSING as BatchJobStatus
}

const fileServiceMock = {
  delete: jest.fn(),
  getUploadStreamDescriptor: jest.fn().mockImplementation(() => {
    return Promise.resolve({
      writeStream: {
        write: (data: string) => {
          outputDataStorage.push(data)
        },
        end: () => void 0
      },
      promise: Promise.resolve(),
      url: 'product-export.csv'
    })
  })
}
const batchJobServiceMock = {
  withTransaction: function () {
    return this
  },
  update: jest.fn().mockImplementation(data => {
    fakeJob = {
      ...fakeJob,
      ...data
    }
    return Promise.resolve(fakeJob)
  }),
  updateStatus: jest.fn().mockImplementation((status) => {
    fakeJob.status = status
    return Promise.resolve(fakeJob)
  }),
  complete: jest.fn().mockImplementation(() => {
    fakeJob.status = BatchJobStatus.COMPLETED
    return Promise.resolve(fakeJob)
  }),
  ready: jest.fn().mockImplementation(() => {
    fakeJob.status = BatchJobStatus.READY
    return Promise.resolve(fakeJob)
  }),
  retrieve: jest.fn().mockImplementation(() => {
    return Promise.resolve(fakeJob)
  })
}
const productServiceMock = {
  withTransaction: function () {
    return this
  },
  list: jest.fn().mockImplementation(() => Promise.resolve(productsToExport)),
  count: jest.fn().mockImplementation(() => Promise.resolve(productsToExport.length)),
  listAndCount: jest.fn().mockImplementation(() => Promise.resolve([productsToExport, productsToExport.length])),
}
const managerMock = MockManager
const productRepositoryMock = {
  ...MockRepository(),
  query: jest.fn().mockImplementation(() => {
    return Promise.resolve([{
      maxOptionsCount: Math.max(...productsToExport.map(p => p.options.length)),
      maxImagesCount: Math.max(...productsToExport.map(p => p.images.length)),
      maxVariantsCount: Math.max(...productsToExport.map(p => p.variants.length)),
      maxMoneyAmountCount: Math.max(...productsToExport.map(p => p.variants.map(v => v.prices.length)).flat()),
    }])
  })
}

describe("Product export strategy", () => {
  const productExportStrategy = new ProductExportStrategy({
    manager: managerMock,
    fileService: fileServiceMock as any,
    batchJobService: batchJobServiceMock as any,
    productService: productServiceMock as any,
    productRepository: productRepositoryMock
  })

  it('should generate the appropriate template', async () => {
    const template = await productExportStrategy.buildTemplate()
    expect(template).toMatch(/.*Product Handle.*/)
    expect(template).toMatch(/.*Product Title.*/)
    expect(template).toMatch(/.*Product Subtitle.*/)
    expect(template).toMatch(/.*Product Description.*/)
    expect(template).toMatch(/.*Product Status.*/)
    expect(template).toMatch(/.*Product Thumbnail.*/)
    expect(template).toMatch(/.*Product Weight.*/)
    expect(template).toMatch(/.*Product Length.*/)
    expect(template).toMatch(/.*Product Width.*/)
    expect(template).toMatch(/.*Product Height.*/)
    expect(template).toMatch(/.*Product HS Code.*/)
    expect(template).toMatch(/.*Product Origin Country.*/)
    expect(template).toMatch(/.*Product Mid Code.*/)
    expect(template).toMatch(/.*Product Material.*/)
    expect(template).toMatch(/.*Product Collection Title.*/)
    expect(template).toMatch(/.*Product Collection Handle.*/)
    expect(template).toMatch(/.*Product Type.*/)
    expect(template).toMatch(/.*Product Tags.*/)
    expect(template).toMatch(/.*Product Discountable.*/)
    expect(template).toMatch(/.*Product External ID.*/)
    expect(template).toMatch(/.*Product Profile Name.*/)
    expect(template).toMatch(/.*Product Profile Type.*/)
    expect(template).toMatch(/.*Product Profile Type.*/)

    expect(template).toMatch(/.*Variant Title.*/)
    expect(template).toMatch(/.*Variant SKU.*/)
    expect(template).toMatch(/.*Variant Barcode.*/)
    expect(template).toMatch(/.*Variant Allow backorder.*/)
    expect(template).toMatch(/.*Variant Manage inventory.*/)
    expect(template).toMatch(/.*Variant Weight.*/)
    expect(template).toMatch(/.*Variant Length.*/)
    expect(template).toMatch(/.*Variant Width.*/)
    expect(template).toMatch(/.*Variant Height.*/)
    expect(template).toMatch(/.*Variant HS Code.*/)
    expect(template).toMatch(/.*Variant Origin Country.*/)
    expect(template).toMatch(/.*Variant Mid Code.*/)
    expect(template).toMatch(/.*Variant Material.*/)

    expect(template).toMatch(/.*Option 1 Name.*/)
    expect(template).toMatch(/.*Option 1 Value.*/)
    expect(template).toMatch(/.*Option 2 Name.*/)
    expect(template).toMatch(/.*Option 2 Value.*/)

    expect(template).toMatch(/.*Price 1 Currency code.*/)
    expect(template).toMatch(/.*Price 1 Region name.*/)
    expect(template).toMatch(/.*Price 1 Amount.*/)

    expect(template).toMatch(/.*Image 1 Url.*/)
  })

  it('should process the batch job and generate the appropriate output', async () => {
    await productExportStrategy.processJob(fakeJob.id)
    expect(outputDataStorage).toMatchSnapshot()
  })
})
