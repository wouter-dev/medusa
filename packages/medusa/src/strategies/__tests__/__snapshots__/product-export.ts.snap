// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product export strategy should process the batch job and generate the appropriate output 1`] = `
Array [
  "Product Handle;Product Title;Product Subtitle;Product Description;Product Status;Product Thumbnail;Product Weight;Product Length;Product Width;Product Height;Product HS Code;Product Origin Country;Product Mid Code;Product Material;Product Collection Title;Product Collection Handle;Product Type;Product Tags;Product Discountable;Product External ID;Product Profile Name;Product Profile Type;Variant Title;Variant SKU;Variant Barcode;Variant Inventory Quantity;Variant Allow backorder;Variant Manage inventory;Variant Weight;Variant Length;Variant Width;Variant Height;Variant HS Code;Variant Origin Country;Variant Mid Code;Variant Material;Option 1 Name;Option 1 Value;Option 2 Name;Option 2 Value;Price 1 Currency code;Price 1 Region name;Price 1 Amount;Image 1 Url;
",
  "test-product-product-1;Test product;;test-product-description-1;draft;;;;;;;;;;Test collection 1;test-collection1;test-type-1;123_1;true;;profile_1;profile_type_1;Test variant;test-sku;test-barcode;10;false;true;;;;;;;;;test-option-1;option 1 value 1;test-option-2;option 2 value 1;usd;;100;test-image.png;
",
  "test-product-product-2;Test product;;test-product-description;draft;;;;;;;;;;Test collection;test-collection2;test-type;123;true;;profile_2;profile_type_2;Test variant;test-sku;test-barcode;10;false;true;;;;;;;;;test-option;Option 1 value 1;;;usd;;100;test-image.png;
",
  "test-product-product-2;Test product;;test-product-description;draft;;;;;;;;;;Test collection;test-collection2;test-type;123;true;;profile_2;profile_type_2;Test variant;test-sku;test-barcode;10;false;true;;;;;;;;;test-option;Option 1 Value 1;;;usd;;100;test-image.png;
",
]
`;
